docs/guide/config.md
# Configuration

::: tip Angles definitions
Photo Sphere Viewer uses a lot of angles for its configuration, most of them can be defined in radians by using a simple number (`3.5`) or in degrees using the "deg" suffix (`'55deg'`).
:::

## Standard options

#### `container` (required)

-   type: `HTMLElement | string`

HTML element which will contain the panorama, or identifier of the element.

```js
container: document.querySelector('.viewer');
container: '.viewer'; // will target [class="viewer"]
container: 'viewer'; // will target [id="viewer"]
```

#### `panorama` (required)

-   type: `*`

Path to the panorama. Must be a single URL for the default equirectangular adapter. Other adapters support other values.

#### `adapter`

-   default: `equirectangular`

Which [adapter](./adapters/) used to load the panorama.

#### `plugins`

-   type: `array`

List of enabled [plugins](../plugins/).

#### `caption`

-   type: `string`

A text displayed in the navbar. If the navbar is disabled, the caption won't be visible.

_**Note:** the content is rendered as HTML. If your content may include untrusted input it should be sanitized before being passed to `caption`._

#### `description`

-   type: `string`

A text displayed in the side panel when the user clicks the "i" button. HTML is allowed.

_**Note:** the content is rendered as HTML. If your content may include untrusted input it should be sanitized before being passed to `description`._

#### `downloadUrl`

-   type: `string`
-   default: `=panorama` for equirectangular panoramas

Define the file which will be downloaded with the `download` button. This is particularly useful for adapters that use multiple files, like the CubemapAdapter or the EquirectangularTilesAdapter.

#### `downloadName`

-   type: `string`
-   default: `=downloadUrl` filename

Overrides the filename when downloading the panorama. This is mostly useful if the panorama is provided as base64.

#### `size`

-   type: `{ width: integer, height: integer }`

The final size of the panorama container. By default the size of `container` is used and is followed when resized.

#### `navbar`

Configuration of the [navbar](./navbar.md).

#### `minFov`

-   type: `integer`
-   default: `30`

Minimal field of view (maximum zoom), between 1 and `maxFov`.

#### `maxFov`

-   type: `integer`
-   default: `90`

Maximal field of view (minimum zoom), between `minFov` and 180.

#### `defaultZoomLvl`

-   type: `integer`
-   default: `50`

Initial zoom level, between 0 (for `maxFov`) and 100 (for `minfov`).

#### `fisheye`

-   type: `boolean | double`
-   default: `false`

Enable fisheye effect with `true` or specify the effect strength (`true` = `1.0`).

<DemoButton href="/demos/basic/fisheye.html"/>

::: warning
This mode can have side-effects on markers rendering and some adapters.
:::

#### `defaultYaw`

-   type: `double | string`
-   default: `0`

Initial horizontal angle, between 0 and 2π.

#### `defaultPitch`

-   type: `double | string`
-   default: `0`

Initial vertical angle, between -π/2 and π/2.

#### `lang`

-   type: `object`
-   default:

```js:line-numbers
lang: {
  zoom: 'Zoom',
  zoomOut: 'Zoom out',
  zoomIn: 'Zoom in',
  moveUp: 'Move up',
  moveDown: 'Move down',
  moveLeft: 'Move left',
  moveRight: 'Move right',
  description: 'Description',
  download: 'Download',
  fullscreen: 'Fullscreen',
  loading: 'Loading...',
  menu: 'Menu',
  close: 'Close',
  twoFingers: 'Use two fingers to navigate',
  ctrlZoom: 'Use ctrl + scroll to zoom the image',
  loadError: 'The panorama cannot be loaded',
  webglError: 'Your browser does not seem to support WebGL',
}
```

Various texts used in the viewer.

#### `loadingImg`

-   type: `string`

Path to an image displayed in the center of the loader.

#### `loadingTxt`

-   type: `string`
-   default: `lang.loading`

Text displayed in the center of the loader, only used if `loadingImg` is not provided.

#### `mousewheel`

-   type: `boolean`
-   default: `true`

Enables zoom with the mouse wheel.

#### `mousemove`

-   type: `boolean`
-   default: `true`

Enables panorama rotation with mouse click+move or with a finger swipe on touch screens.

#### `keyboard`

-   type: `boolean | 'fullscreen' | 'always'`
-   default: `'fullscreen'` (same as `true`)

Enables keyboard controls when in fullscreen or always. The different keys can be configured with [`keyboardActions`](#keyboardactions).

::: warning
Keys are listened globally to the page, and thus can be in conflict with other components if configured to `'always'`.
:::

#### `mousewheelCtrlKey`

-   type: `boolean`
-   default: `false`

Requires to use the ctrl key to zoom the panorama. This allows to scroll on the page without interfering with the viewer. If enabled, an overlay asking the user to use ctrl + scroll is displayed when ctrl key is not pressed.

#### `touchmoveTwoFingers`

-   type: `boolean`
-   default: `false`

Requires two fingers to rotate the panorama. This allows standard touch-scroll navigation in the page containing the viewer. If enabled, an overlay asking the user to use two fingers is displayed when only one finger is detected.

## Advanced options

#### `sphereCorrection`

-   type: `{ pan: double | string, tilt: double | string, roll: double | string }`
-   default: `{ pan:0, tilt:0, roll: 0 }`

Allows to fix the panorama orientation.

**Note:** if the XMP data contains pose heading/pitch/roll data, they will be applied before `sphereCorrection`.

![pan-tilt-toll](/images/pan-tilt-roll.png)

#### `panoData`

-   type: `PanoData | function<Image, PanoData, PanoData>`

Overrides XMP data found in the panorama file.
All parameters are optional.

```js:line-numbers
panoData: {
  fullWidth: 6000,
  fullHeight: 3000, // optional
  croppedWidth: 4000, // optional
  croppedHeight: 2000, // optional
  croppedX: 1000,
  croppedY: 500,
}
```

It can also be a function to dynamically compute the cropping config depending on the loaded image (note that a [default setting](./adapters/equirectangular.md#default-parameters) is already applied when no data is found).

```js:line-numbers
panoData: (image, xmpData) => ({
    fullWidth: image.width,
    fullHeight: Math.round(image.width / 2),
    croppedWidth: image.width,
    croppedHeight: image.height,
    croppedX: 0,
    croppedY: Math.round((image.width / 2 - image.height) / 2),
});
```

#### `defaultTransition`

-   type: `TransitionOptions`
-   default: `{ speed: 1500, rotation: true, effect: "fade" }`

Configures the default transition between panoramas. All parameters can be changed when calling `setPanorama()` method.

Possible `effect` are `fade, black, white`.  
The `speed` is either a duration in milliseconds or a string containing the speed in revolutions per minute (`2rpm`).


<DemoButton href="/demos/basic/transition.html"></DemoButton>

#### `moveSpeed`

-   type: `double`
-   default `1`

Speed multiplicator for panorama moves. Used for click move, touch move and navbar buttons.

#### `zoomSpeed`

-   type: `double`
-   default `1`

Speed multiplicator for panorama zooms. Used for mouse wheel, touch pinch and navbar buttons.

#### `moveInertia`

-   type: `boolean | number`
-   default: `0.8`

Applies damping to the camera movement, higher value mean stronger damping (`true` is default damping factor, `false` is not damping).

#### `requestHeaders`

-   type: `object | function<string, object>`

Sets the HTTP headers when loading the images files.

```js
requestHeaders: {
    header: 'value',
}
```

It can also be a function to dynamically set the request headers before every call. This can be useful when adding a Bearer, which is temporarily valid, to the Authorization header.

```js
requestHeaders: (url) => ({
    header: 'value',
});
```

#### `withCredentials`

-   type: `boolean | function<string, boolean>`
-   default: `false`

Use credentials for HTTP requests.

It can also be a function to dynamically change the option before every call.

```js
withCredentials: (url) => !url.includes('amazonaws');
```

#### `keyboardActions`

-   type: `object`
-   default:

```js:line-numbers
keyboardActions: {
  'ArrowUp': 'ROTATE_UP',
  'ArrowDown': 'ROTATE_DOWN',
  'ArrowRight': 'ROTATE_RIGHT',
  'ArrowLeft': 'ROTATE_LEFT',
  'PageUp': 'ZOOM_IN',
  'PageDown': 'ZOOM_OUT',
  '+': 'ZOOM_IN',
  '-': 'ZOOM_OUT',
}
```

Configure keyboard actions. It is a map defining key code->action. The key can include modifiers like `Ctrl`, `Shift`, `Alt` or `Meta` separated with a `+` symbol.

::: tip
Use [Key.js](https://keyjs.dev/) wbesite to determine exact combination of a keystroke.

_Note:_ Use `Plus` instead of `+` when configuring the plus key with a modifier, for example `Shift+Plus`.
:::

You can also configure an arbitrary callback to any key, it receives the viewer itself and the original keyboard event as parameters.

```js:line-numbers
import { DEFAULTS } from '@photo-sphere-viewer/core';

keyboardActions: {
    ...DEFAULTS.keyboardActions,
    'h': (viewer, evt) => {
        // do something when H is pressed
    },
    'Ctrl+1': (viewer) => {
        // do something when Ctrl + 1 is pressed
    },
},
```

<DemoButton href="/demos/advanced/keyboard-actions.html"/>

::: warning
Keyboard actions will only be available in fullscreen by default, this can be changed with the [`keyboard` option](#keyboard).
:::

#### `canvasBackground`

-   type: `string`
-   default: `#000`

Background of the canvas, which will be visible when using cropped panoramas. Can be any valid CSS `background` value.

#### `rendererParameters`

-   type: [`WebGLRendererParameters`](https://threejs.org/docs/#api/en/renderers/WebGLRenderer)
-   default: `{ alpha: true, antialias: true }`

Configuration of the ThreeJS WebGLRenderer.

## Cache

Photo Sphere Viewer contains a cache system to save resources when switching back and forth between multiple panoramas.
This cache is global and shared across viewers (_note:_ is not related to ThreeJS Cache, which should not be enabled).

To get the cache instance, import it from `@photo-sphere-viewer/core`, then you can configure it.

```js
import { Cache } from '@photo-sphere-viewer/core';

Cache.enabled = false;
Cache.ttl = 300;
Cache.maxItems = 3;
```

#### `enabled`

-   type: `boolean`
-   default: `true`

Allows to completely disable the cache.

#### `ttl`

-   type: `number`
-   default: `600`

Maximum retention duration in minutes.

#### `maxItems`

-   type: `number`
-   default: `10`

Maximum number of items stored in the case.

_Note:_ the actual number of files will be greater with cubemap and tiles adapters.


docs/plugins/settings.md
# SettingsPlugin <Badge text="Styles"/>

<Badges module="settings-plugin"/>

::: module
<ApiButton page="modules/SettingsPlugin.html"/>
This plugin does nothing on it's own but is required by other plugins.

This plugin is available in the [@photo-sphere-viewer/settings-plugin](https://www.npmjs.com/package/@photo-sphere-viewer/settings-plugin) package.
:::

## Usage

Once enabled the plugin will add a new "Settings" button which can be populated by other plugins.

```js:line-numbers
import { SettingsPlugin } from '@photo-sphere-viewer/settings-plugin';

const viewer = new Viewer({
    plugins: [
        SettingsPlugin,
    ],
});
```

## Example

The following example manually adds two settings.

::: code-demo

```yaml
title: PSV Settings Demo
packages:
    - name: settings-plugin
      style: true
```

<<< ./demos-src/settings.js{js:line-numbers}

:::

## Adding a setting

Registering a new setting is done by calling the `addSetting` on the plugin. There are currently two types of setting.

### Toggle setting

This a setting which has only two values : `true` and `false`. It is required to provide the `active(): boolean` and `toggle(): void` functions.

```js:line-numbers
let enabled = false;

settings.addSetting({
    id: 'custom-toggle-setting',
    label: 'Toggle setting',
    type: 'toggle',
    active: () => enabled,
    toggle: () => (enabled = !enabled),
});
```

### Options setting

This is a setting which has multiple available values (or options). It is required to provide the `current(): string`, `options(): Options[]` and `apply(option: string): void` functions.

```js:line-numbers
let currentOption = 'A';

settings.addSetting({
    id: 'custom-options-setting',
    label: 'Options setting',
    type: 'options',
    options: () => [
        { id: 'A', label: 'Option A' },
        { id: 'B', label: 'Option B' },
    ],
    current: () => currentOption,
    apply: (option) => (currentOption = option),
});
```

For translation purposes, both `label` can be a key in the main [`lang`](../guide/config.md#lang) object.

## Button badge

A setting can also have a `badge` function, which return value will be used as a badge on the settings button itself. **Only one setting can declare a badge.**

```js:line-numbers
settings.addSetting({
  ...,
  badge: () => currentOption,
});
```

## Configuration

#### `persist`

-   type: `boolean`
-   default: `false`
-   updatable: no

Should the settings be persisted. The persistence storage can be configured.

#### `storage`

-   type:

```ts:line-numbers
{
  get(settingId: string): boolean | string | Promise<boolean | string>;
  set(settingId: string, value: boolean | string);
}
```

-   default: LocalStorage with key `psvSettings`
-   updatable: no

Custom storage solution, for example LocalForage, NgRx, HTTP service, etc.

#### `lang`

-   type: `object`
-   default:

```js
lang: {
    settings: 'Settings',
}
```

_Note: this option is not part of the plugin but is merged with the main [`lang`](../guide/config.md#lang) object._

## Events

#### `setting-changed(settingId, settingValue)`

Triggered when the resolution is changed.

```js:line-numbers
settingsPlugin.addEventListener('setting-changed', ({ settingId, settingValue }) => {
    console.log(`${settingId}: ${settingValue}`);
});
```

## Buttons

This plugin adds buttons to the default navbar:

-   `settings` allows to open the settings panel

If you use a [custom navbar](../guide/navbar.md) you will need to manually add the buttons to the list.

## SCSS variables

Check the main [style](../guide/style.md) documentation if you want to customize the settings style.
The full list of SCSS variables can be found [in the source code](https://github.com/mistic100/Photo-Sphere-Viewer/blob/main/packages/settings-plugin/src/styles/_vars.scss).


docs/changelog.md
# Changelog

<Changelog/>


src/data/config.js
import { MathUtils } from 'three';
import { AbstractAdapter } from '../adapters/AbstractAdapter';
import { EquirectangularAdapter } from '../adapters/equirectangular';
import { AbstractPlugin } from '../plugins/AbstractPlugin';
import { PSVError } from '../PSVError';
import { clone, deepmerge, each, isNil, logWarn, parseAngle, parseSpeed, pluginInterop } from '../utils';
import { ACTIONS, KEY_CODES } from './constants';

/**
 * @summary Default options
 * @type {PSV.Options}
 * @memberOf PSV
 * @constant
 */
export const DEFAULTS = {
  panorama           : null,
  overlay            : null,
  overlayOpacity     : 1,
  container          : null,
  adapter            : null,
  plugins            : [],
  caption            : null,
  description        : null,
  downloadUrl        : null,
  loadingImg         : null,
  loadingTxt         : 'Loading...',
  size               : null,
  fisheye            : false,
  minFov             : 30,
  maxFov             : 90,
  defaultZoomLvl     : 50,
  defaultLong        : 0,
  defaultLat         : 0,
  sphereCorrection   : null,
  moveSpeed          : 1,
  zoomSpeed          : 1,
  autorotateDelay    : null,
  autorotateIdle     : false,
  autorotateSpeed    : '2rpm',
  autorotateLat      : null,
  autorotateZoomLvl  : null,
  moveInertia        : true,
  mousewheel         : true,
  mousemove          : true,
  mousewheelCtrlKey  : false,
  touchmoveTwoFingers: false,
  useXmpData         : true,
  panoData           : null,
  requestHeaders     : null,
  canvasBackground   : '#000',
  withCredentials    : false,
  navbar             : [
    'autorotate',
    'zoom',
    'move',
    'download',
    'description',
    'caption',
    'fullscreen',
  ],
  lang               : {
    autorotate: 'Automatic rotation',
    zoom      : 'Zoom',
    zoomOut   : 'Zoom out',
    zoomIn    : 'Zoom in',
    move      : 'Move',
    download  : 'Download',
    fullscreen: 'Fullscreen',
    menu      : 'Menu',
    twoFingers: 'Use two fingers to navigate',
    ctrlZoom  : 'Use ctrl + scroll to zoom the image',
    loadError : 'The panorama can\'t be loaded',
  },
  keyboard           : {
    [KEY_CODES.ArrowUp]   : ACTIONS.ROTATE_LAT_UP,
    [KEY_CODES.ArrowDown] : ACTIONS.ROTATE_LAT_DOWN,
    [KEY_CODES.ArrowRight]: ACTIONS.ROTATE_LONG_RIGHT,
    [KEY_CODES.ArrowLeft] : ACTIONS.ROTATE_LONG_LEFT,
    [KEY_CODES.PageUp]    : ACTIONS.ZOOM_IN,
    [KEY_CODES.PageDown]  : ACTIONS.ZOOM_OUT,
    [KEY_CODES.Plus]      : ACTIONS.ZOOM_IN,
    [KEY_CODES.Minus]     : ACTIONS.ZOOM_OUT,
    [KEY_CODES.Space]     : ACTIONS.TOGGLE_AUTOROTATE,
  },
};

/**
 * @summary List of unmodifiable options and their error messages
 * @private
 */
export const READONLY_OPTIONS = {
  panorama : 'Use setPanorama method to change the panorama',
  panoData : 'Use setPanorama method to change the panorama',
  container: 'Cannot change viewer container',
  adapter  : 'Cannot change adapter',
  plugins  : 'Cannot change plugins',
};

/**
 * @summary List of deprecated options and their warning messages
 * @private
 */
export const DEPRECATED_OPTIONS = {
  captureCursor: 'captureCursor is deprecated',
};

/**
 * @summary Parsers/validators for each option
 * @private
 */
export const CONFIG_PARSERS = {
  container        : (container) => {
    if (!container) {
      throw new PSVError('No value given for container.');
    }
    return container;
  },
  adapter          : (adapter) => {
    if (!adapter) {
      adapter = [EquirectangularAdapter];
    }
    else if (Array.isArray(adapter)) {
      adapter = [pluginInterop(adapter[0], AbstractAdapter), adapter[1]];
    }
    else {
      adapter = [pluginInterop(adapter, AbstractAdapter)];
    }
    if (!adapter[0]) {
      throw new PSVError('Un undefined value with given for adapter.');
    }
    return adapter;
  },
  overlayOpacity   : (overlayOpacity) => {
    return MathUtils.clamp(overlayOpacity, 0, 1);
  },
  defaultLong      : (defaultLong) => {
    // defaultLat is between 0 and PI
    return parseAngle(defaultLong);
  },
  defaultLat       : (defaultLat) => {
    // defaultLat is between -PI/2 and PI/2
    return parseAngle(defaultLat, true);
  },
  defaultZoomLvl   : (defaultZoomLvl) => {
    return MathUtils.clamp(defaultZoomLvl, 0, 100);
  },
  minFov           : (minFov, config) => {
    // minFov and maxFov must be ordered
    if (config.maxFov < minFov) {
      logWarn('maxFov cannot be lower than minFov');
      minFov = config.maxFov;
    }
    // minFov between 1 and 179
    return MathUtils.clamp(minFov, 1, 179);
  },
  maxFov           : (maxFov, config) => {
    // minFov and maxFov must be ordered
    if (maxFov < config.minFov) {
      maxFov = config.minFov;
    }
    // maxFov between 1 and 179
    return MathUtils.clamp(maxFov, 1, 179);
  },
  lang             : (lang) => {
    if (Array.isArray(lang.twoFingers)) {
      logWarn('lang.twoFingers must not be an array');
      lang.twoFingers = lang.twoFingers[0];
    }
    return {
      ...DEFAULTS.lang,
      ...lang,
    };
  },
  keyboard         : (keyboard) => {
    // keyboard=true becomes the default map
    if (keyboard === true) {
      return clone(DEFAULTS.keyboard);
    }
    return keyboard;
  },
  autorotateLat    : (autorotateLat, config) => {
    // default autorotateLat is defaultLat
    if (autorotateLat === null) {
      return parseAngle(config.defaultLat, true);
    }
    // autorotateLat is between -PI/2 and PI/2
    else {
      return parseAngle(autorotateLat, true);
    }
  },
  autorotateZoomLvl: (autorotateZoomLvl) => {
    if (isNil(autorotateZoomLvl)) {
      return null;
    }
    else {
      return MathUtils.clamp(autorotateZoomLvl, 0, 100);
    }
  },
  autorotateSpeed  : (autorotateSpeed) => {
    return parseSpeed(autorotateSpeed);
  },
  autorotateIdle   : (autorotateIdle, config) => {
    if (autorotateIdle && isNil(config.autorotateDelay)) {
      logWarn('autorotateIdle requires a non null autorotateDelay');
      return false;
    }
    return autorotateIdle;
  },
  fisheye          : (fisheye) => {
    // translate boolean fisheye to amount
    if (fisheye === true) {
      return 1;
    }
    else if (fisheye === false) {
      return 0;
    }
    return fisheye;
  },
  plugins          : (plugins) => {
    return plugins
      .map((plugin) => {
        if (Array.isArray(plugin)) {
          plugin = [pluginInterop(plugin[0], AbstractPlugin), plugin[1]];
        }
        else {
          plugin = [pluginInterop(plugin, AbstractPlugin)];
        }
        if (!plugin[0]) {
          throw new PSVError('Un undefined value was given for plugins.');
        }
        return plugin;
      });
  },
};

/**
 * @summary Merge user config with default config and performs validation
 * @param {PSV.Options} options
 * @returns {PSV.Options}
 * @memberOf PSV
 * @private
 */
export function getConfig(options) {
  const tempConfig = clone(DEFAULTS);
  deepmerge(tempConfig, options);

  const config = {};

  each(tempConfig, (value, key) => {
    if (DEPRECATED_OPTIONS[key]) {
      logWarn(DEPRECATED_OPTIONS[key]);
      return;
    }

    if (!Object.prototype.hasOwnProperty.call(DEFAULTS, key)) {
      throw new PSVError(`Unknown option ${key}`);
    }

    if (CONFIG_PARSERS[key]) {
      config[key] = CONFIG_PARSERS[key](value, tempConfig);
    }
    else {
      config[key] = value;
    }
  });

  return config;
}
